pageserver/src/tenant/storage_layer.rs

```diff
@@ -81,6 +88,103 @@ pub enum ValueReconstructResult {
     Missing,
 }
 
+#[derive(Default)]
+pub struct LayerAccessStats(RwLock<LayerAccessStatsInner>);
+
+#[derive(Clone)]
+struct LayerAccessStatFullDetails {
+    when: SystemTime,
+    task_kind: TaskKind,
+    access_kind: LayerAccessKind,
+}
+
+#[derive(Default)]
+struct LayerAccessStatsInner {
+    first_access: Option<LayerAccessStatFullDetails>,
+    count_by_access_kind: EnumMap<LayerAccessKind, u64>,
+    task_kind_flag: EnumSet<TaskKind>,
+    last_accesses: VecDeque<LayerAccessStatFullDetails>,
+}
+
+#[derive(Clone, Copy, strum_macros::EnumString)]
+pub enum LayerAccessStatsReset {
+    JustTaskKindFlags,
+    AllStats,
+}
+
+impl LayerAccessStatFullDetails {
+    fn to_api_model(&self) -> pageserver_api::models::LayerAccessStatFullDetails {
+        let Self {
+            when,
+            task_kind,
+            access_kind,
+        } = self;
+        pageserver_api::models::LayerAccessStatFullDetails {
+            when_millis_since_epoch: when
+                .duration_since(UNIX_EPOCH)
+                .expect("better to die in this unlikely case than report false stats")
+                .as_millis(),
+            task_kind: task_kind.into(), // into static str, powered by strum_macros
+            access_kind: *access_kind,
+        }
+    }
+}
+
+pub static LAYER_ACCESS_STATS_KILLSWITCH: AtomicBool = AtomicBool::new(false);
+
+impl LayerAccessStats {
+    fn record_access(&self, access_kind: LayerAccessKind, task_kind: TaskKind) {
+        if LAYER_ACCESS_STATS_KILLSWITCH.load(atomic::Ordering::SeqCst) {
+            return;
+        }
+        let mut inner = self.0.write().unwrap();
+        let this_access = LayerAccessStatFullDetails {
+            when: SystemTime::now(),
+            task_kind,
+            access_kind,
+        };
+        inner
+            .first_access
+            .get_or_insert_with(|| this_access.clone());
+        inner.count_by_access_kind[access_kind] += 1;
+        inner.task_kind_flag |= task_kind;
+        inner.last_accesses.push_front(this_access);
+        inner.last_accesses.truncate(10); // TODO make runtime-configurable
```

Comment:
```suggestion
        while inner.last_accesses.len() >= LIMIT {
            inner.last_accesses.pop_back();
        }
        inner.last_accesses.push_front(this_access);
```

Though I guess with limit == 10 this doesn't matter, but in general, push could reallocate, so the limit needs to be checked beforehand. In practice, as limit == 10 is not a power of 2, the capacity of the vecdeque will be 16, though this could be relaxed in future std versions.

Would also prefer to see this type have a custom `impl Default` which would specify `VecDeque::with_capacity(LIMIT)`.